ArduPlane/navigation.cpp

```diff
@@ -152,19 +156,47 @@ void Plane::calc_airspeed_errors()
 
 void Plane::calc_gndspeed_undershoot()
 {
- 	// Use the component of ground speed in the forward direction
-	// This prevents flyaway if wind takes plane backwards
-    if (gps.status() >= AP_GPS::GPS_OK_FIX_2D) {
-	    Vector2f gndVel = ahrs.groundspeed_vector();
-		const Matrix3f &rotMat = ahrs.get_rotation_body_to_ned();
-		Vector2f yawVect = Vector2f(rotMat.a.x,rotMat.b.x);
-        if (!yawVect.is_zero()) {
-            yawVect.normalize();
-            float gndSpdFwd = yawVect * gndVel;
-            groundspeed_undershoot = (aparm.min_gndspeed_cm > 0) ? (aparm.min_gndspeed_cm - gndSpdFwd*100) : 0;
-        }
+
+    // Use the component of ground speed in the forward direction
+    // This prevents flyaway if wind takes plane backwards
+    const Matrix3f &rotMat = ahrs.get_rotation_body_to_ned();
+    Vector2f yawVect = Vector2f(rotMat.a.x,rotMat.b.x);
+
+    // Use integrator based gndspd_undershoot controller
+    if (gps.status() >= AP_GPS::GPS_OK_FIX_2D &&
+        is_flying() && !yawVect.is_zero() &&
```

Comment:
The wind-up protection is a must, but I am as to the preferred implementation. There is an alternative heuristic at the end of Plane:stabilize() that does reset integrators, but it is not completely bulletproof neither e.g it could fail in case of strong wind on approach, or when flying in a neighboring valley around the launch point altitude. Nevertheless, while, those are real life scenarios I didn't catch any actual integrator reset in flight.

Reply:
Perhaps, using the throttle_suppressed flag would be better?